### ISSUE TITLE
Terminal setting are initialized asynchronously after container is already up and running

### ISSUE THREAD

**Issue by iltish** [docker/docker#28009]
Looks like starting docker container with **-ti** sets terminal settings while processes inside container are already up and running. This causes some programs (like **rlwrap**) to fail due to the uninitialized terminal settings.

If you put **sleep** even for a very short period of time in front of the command, terminal settings become initialized properly.

This issue was reproduced on Ubuntu and MacOS. However, since this is a race condition, it depends on system speed and current load.

Configuration:
```
>uname -a
Linux ldev1 4.4.0-45-generic #66-Ubuntu SMP Wed Oct 19 14:12:37 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
>docker -v
Docker version 1.12.1, build 23cf638
```

**rlwrap** works only after sleep command:
```
>docker run -ti ubuntu_with_rlwrap bash -c "rlwrap ls"
rlwrap: error: My terminal reports width=0 (is it emacs?)  I can't handle this, sorry!
>docker run -ti ubuntu_with_rlwrap bash -c "sleep 0.0001;rlwrap ls"
bin  boot  dev  etc  home  lib  lib64  media  mnt  opt  proc  root  run  sbin  srv  sys  tmp  usr  var
```

Running **stty** 3 times in a row, with only the last run showing correct terminal setting:
```
>docker run -ti ubuntu bash -c "stty -a;stty -a;sleep 0.01;stty -a"
speed 38400 baud; rows 0; columns 0; line = 0;
intr = ^C; quit = ^\; erase = ^?; kill = ^U; eof = ^D; eol = <undef>;
eol2 = <undef>; swtch = <undef>; start = ^Q; stop = ^S; susp = ^Z; rprnt = ^R;
werase = ^W; lnext = ^V; discard = ^O; min = 1; time = 0;
-parenb -parodd -cmspar cs8 -hupcl -cstopb cread -clocal -crtscts
-ignbrk -brkint -ignpar -parmrk -inpck -istrip -inlcr -igncr icrnl ixon -ixoff
-iuclc -ixany -imaxbel -iutf8
opost -olcuc -ocrnl onlcr -onocr -onlret -ofill -ofdel nl0 cr0 tab0 bs0 vt0 ff0
isig icanon iexten echo echoe echok -echonl -noflsh -xcase -tostop -echoprt
echoctl echoke -flusho -extproc
speed 38400 baud; rows 0; columns 0; line = 0;
intr = ^C; quit = ^\; erase = ^?; kill = ^U; eof = ^D; eol = <undef>;
eol2 = <undef>; swtch = <undef>; start = ^Q; stop = ^S; susp = ^Z; rprnt = ^R;
werase = ^W; lnext = ^V; discard = ^O; min = 1; time = 0;
-parenb -parodd -cmspar cs8 -hupcl -cstopb cread -clocal -crtscts
-ignbrk -brkint -ignpar -parmrk -inpck -istrip -inlcr -igncr icrnl ixon -ixoff
-iuclc -ixany -imaxbel -iutf8
opost -olcuc -ocrnl onlcr -onocr -onlret -ofill -ofdel nl0 cr0 tab0 bs0 vt0 ff0
isig icanon iexten echo echoe echok -echonl -noflsh -xcase -tostop -echoprt
echoctl echoke -flusho -extproc
speed 38400 baud; rows 67; columns 198; line = 0;
intr = ^C; quit = ^\; erase = ^?; kill = ^U; eof = ^D; eol = <undef>; eol2 = <undef>; swtch = <undef>; start = ^Q; stop = ^S; susp = ^Z; rprnt = ^R; werase = ^W; lnext = ^V; discard = ^O;
min = 1; time = 0;
-parenb -parodd -cmspar cs8 -hupcl -cstopb cread -clocal -crtscts
-ignbrk -brkint -ignpar -parmrk -inpck -istrip -inlcr -igncr icrnl ixon -ixoff -iuclc -ixany -imaxbel -iutf8
opost -olcuc -ocrnl onlcr -onocr -onlret -ofill -ofdel nl0 cr0 tab0 bs0 vt0 ff0
isig icanon iexten echo echoe echok -echonl -noflsh -xcase -tostop -echoprt echoctl echoke -flusho -extproc

**Comment by thaJeztah** [comment]
Looks like this is a duplicate of https://github.com/docker/docker/issues/25450, and related https://github.com/docker/docker/issues/10341

I'll close this issue to keep the discussion on a single place, but let me know if you think it's not a duplicate

**Comment by tianon** [comment]
Either this isn't a duplicate or it's not actually fixed -- on the latest 18.09.0 release, I'm still seeing this `rlwrap` error (`rlwrap: error: My terminal reports width=0 (is it emacs?)  I can't handle this, sorry!`) and `tput cols`+`tput lines` via direct `docker run -it ... tput ...` reports 80x24 no matter my source terminal size. :disappointed:

**Comment by tianon** [comment]
(Is that really https://github.com/moby/moby/issues/10341?  I thought the initial terminal size issues should've been fixed already? :disappointed:)
